### PR TITLE
[RLlib] Add DTTorchPolicy

### PR DESCRIPTION
**Pull request by charlesjsun**
## Why are these changes needed?

Add DTTorchPolicy for DT.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/dt/dt_torch_policy.py [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+
+    @override(TorchPolicyV2)
+    def make_model_and_action_dist(
+        self,
+    ) -> Tuple[ModelV2, Type[TorchDistributionWrapper]]:
+        # Model
+        model_config = self.config["model"]
+        # TODO: make these better with better AlgorithmConfig options.
+        model_config.update(
+            dict(
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/dt_torch_policy.py:78]: **TODO**: you can pass kwargs directly to dict.update()

File: rllib/algorithms/dt/dt_torch_policy.py [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+        input_dict = tree.map_structure(convert_to_numpy, input_dict)
+        # convert everything else to numpy as well
+        action, reward, next_obs, extra = convert_to_numpy(
+            (action, reward, next_obs, extra)
+        )
+
+        # Shift observations
+        input_dict[SampleBatch.OBS] = np.concatenate(
+            [
+                input_dict[SampleBatch.OBS][1:],
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/dt_torch_policy.py:242]: should this be `[1:]` or `[-self.max_seq_len + 1:]`? Do you want to check the dimensions inside this method? e.g. input_dict['obs'] has the right number of timesteps?

File: rllib/algorithms/dt/dt_torch_policy.py [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+                    RETURNS_TO_GO: [batch_size, max_seq_len - 1],
+                    REWARDS: [batch_size],
+                    TIMESTEPS: [batch_size, max_seq_len - 1],
+                }
+            state_batches: RNN states (not used).
+            kwargs: forward compatibility (unused) kwargs.
+
+        Returns:
+            A tuple of: (
+                batched input to the action distribution clas,
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/dt_torch_policy.py:310]: typo on clas -> class

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 2 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py:242]: It should be [1:] because we're doing a sliding window, essentially removing the oldest and adding the newer one. It would be good to add a check for the size of the arrays though.

File: rllib/algorithms/dt/dt_torch_policy.py [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+        # -(max_seq_len-1):-1), and subtracts the last reward from it.
+        rtg = obs_batch[SampleBatch.RETURNS_TO_GO]
+        last_rtg = rtg[:, -1]
+        last_reward = obs_batch[SampleBatch.REWARDS]
+        updated_rtg = last_rtg - last_reward
+        # initial_rtg simply is filled with target_return.
+        # These two are both only for the current timestep.
+        initial_rtg = torch.full(
+            (batch_size, 1),
+            fill_value=self.config["target_return"],
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/dt_torch_policy.py:365]: can we create a method in this policy `get_initial_return(shape, device, dtype)` which returns a tensor of correct shape for the the initial rtg? By default the implementation will be what you have now if target_return is specified in config. Otherwise raise an error. This way user can implement their own logic later (say they want to evaluate their policy across a range of rtg values? not just a single one.) They can then override the method to return a random set of rtg values.

Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py]: added

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 2 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py:255]: added the check

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 1 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py]: resolved

File: rllib/algorithms/dt/dt_torch_policy.py [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+
+        losses = []
+
+        # action losses
+        if isinstance(self.action_space, Discrete):
+            action_loss = self._masked_cross_entropy_loss(
+                preds[SampleBatch.ACTIONS], targets[SampleBatch.ACTIONS], masks
+            )
+        elif isinstance(self.action_space, Box):
+            action_loss = self._masked_mse_loss(
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/dt_torch_policy.py:493]: You have a lot of code repetition here. Here is my suggested update:
```
# in loss() call this 
loss = self._masked_loss(preds, targets, masks)
# fn definition
def _masked_loss(preds, targets, masks):
     for key in targets:
           assert key not in preds, 'for target {key} there is no prediction named from the output of the model'
           if self._is_discrete(key):
                loss += self.config[f'loss_coef_{key}'] * self._masked_xent_loss(preds[key], targets[key], masks)
           else:
                loss += self.config[f'loss_coef_{key}'] * self._masked_mse_loss(preds[key], targets[key], masks)

     return loss

def _is_discrete(self, key):
     return key == SampleBatch.Action and self.action_space.is_discrete
```

Comment (kouroshHakha) [reply rllib/algorithms/dt/dt_torch_policy.py:493]: and user can use the coefs to basically tell which output heads they should regress over.

File: rllib/algorithms/dt/tests/test_dt_policy.py [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+import ray
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.framework import try_import_tf, try_import_torch
+from rllib.algorithms.dt.dt_torch_policy import DTTorchPolicy
+
+tf1, tf, tfv = try_import_tf()
+torch, nn = try_import_torch()
+
+
+def _assert_outputs_equal(outputs):
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/tests/test_dt_policy.py:16]: These function are not used ??

File: rllib/algorithms/dt/tests/test_dt_policy.py [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+            # Run loss on a widely different batch and make sure the loss is different.
+            batch3 = SampleBatch(
+                {
+                    SampleBatch.OBS: np.array(
+                        [
+                            [
+                                [1.0, 1.0, -20.0],
+                                [0.1, 10.0, 12.0],
+                                [0.0, 1.0, 2.0],
+                                [3.0, 40.0, 5.0],
```
Comment (kouroshHakha) [inline rllib/algorithms/dt/tests/test_dt_policy.py:409]: 40 -> 4?

Comment (kouroshHakha) [reply rllib/algorithms/dt/tests/test_dt_policy.py:409]: row 3 col 1

**Review by kouroshHakha — COMMENTED** [review]
Thanks for the good documentation which made it easy to follow along. Please address the remaining comments.

File: rllib/algorithms/dt/tests/test_dt_policy.py (hunk 7 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/tests/test_dt_policy.py:409]: They are just random values

File: rllib/algorithms/dt/dt_torch_policy.py [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+
+        # policy forward and get prediction
+        model_out, _ = self.model(train_batch)  # noop, just returns obs.
+        preds = self.model.get_prediction(model_out, train_batch)
+
+        # get the regression targets
+        targets = self.model.get_targets(model_out, train_batch)
+
+        # get the attention masks for masked-loss
+        masks = train_batch[SampleBatch.ATTENTION_MASKS]
```
[the remark this answers is not in the document]
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py:483]: Yes they should be the same. We mask the output of the inputs that are padded.

File: rllib/algorithms/dt/dt_torch_policy.py [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+
+        # Run inference on model
+        model_out, _ = model(obs_batch)  # noop, just returns obs.
+        preds = self.model.get_prediction(model_out, obs_batch)
+        pred_action = preds[SampleBatch.ACTIONS][:, -1]
+
+        return pred_action, self.dist_class, []
+
+    @override(TorchPolicyV2)
+    def extra_action_out(
```
[the remark this answers is not in the document]
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py:429]: I don't like it either. If you think copying some code is worth it then I'll override it. But it's actually `_compute_action_helper` we have to override and not `compute_actions_from_input_dict`

File: rllib/algorithms/dt/tests/test_dt_policy.py (hunk 6 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/tests/test_dt_policy.py]: resolved

File: rllib/algorithms/dt/tests/test_dt_policy.py (hunk 7 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/tests/test_dt_policy.py]: made them more different than batch1

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 5 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py]: done

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/algorithms/dt/dt_torch_policy.py:483]: kk

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 5 above)
Comment (kouroshHakha) [reply rllib/algorithms/dt/dt_torch_policy.py]: You should create a unit-tests that tests whether setting the loss_coefficients properly is actually effective and not buggy.

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py]: I overrided compute_actions_from_input_dict

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 5 above)
Comment (charlesjsun) [reply rllib/algorithms/dt/dt_torch_policy.py]: Added the tests

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/algorithms/dt/dt_torch_policy.py]: nice!

File: rllib/algorithms/dt/dt_torch_policy.py (hunk 5 above)
Comment (kouroshHakha) [reply rllib/algorithms/dt/dt_torch_policy.py]: awesome!

**Review by kouroshHakha — APPROVED** [review]
@charlesjsun Great work in addressing everything man! @richardliaw Can we get this in the merge queue once all the tests pass again?
